File: openstack-worker/app/routes/vms.py
from fastapi import APIRouter
from pydantic import BaseModel
from app.services import vm_service

router = APIRouter(prefix="/api/v1")

class CreateVMRequest(BaseModel):
    name: str
    image_id: str
    flavor_id: str
    network_id: str

@router.get("/vms")
def list_vms():
    return vm_service.list_vms_service()

@router.get("/vms/{vm_id}")
def get_vm_by_id(vm_id: str):
    return vm_service.get_vm_by_id_service(vm_id)

@router.post("/vms")
def create_vm(body: CreateVMRequest):
    return vm_service.create_vm_service(body.name, body.image_id, body.flavor_id, body.network_id)

@router.delete("/vms/{vm_id}")
def delete_vm(vm_id: str):
    vm_service.delete_vm_service(vm_id)
    return {"message": "VM supprimée"}

@router.post("/vms/{vm_id}/start")
def start_vm(vm_id: str):
    vm_service.start_vm_service(vm_id)
    return {"message": "VM démarrée"}

@router.post("/vms/{vm_id}/stop")
def stop_vm(vm_id: str):
    vm_service.stop_vm_service(vm_id)
    return {"message": "VM arrêtée"}

@router.post("/vms/{vm_id}/reboot")
def reboot_vm(vm_id: str):
    vm_service.reboot_vm_service(vm_id)
    return {"message": "VM redémarrée"}

@router.get("/vms/{vm_id}/volumes")
def get_volumes_by_vm(vm_id: str):
    return vm_service.get_volumes_by_vm_service(vm_id)

File: apps/api/src/routes/vms.ts
import type { FastifyInstance } from 'fastify'
import * as workerService from '../services/openstack.js'
import authenticate from '../plugins/authenticate.js'
import { validateQuery, providerIdQuerySchema } from '../utils/validation.js'

export async function vmRoutes(server: FastifyInstance) {
  server.addHook('preHandler', authenticate)

  server.get('/vms', async (req, reply) => {
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.getVms(providerId)
    return reply.send(data)
  })

  server.get('/vms/:id', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.getVmById(providerId, id)
    return reply.send(data)
  })

  server.post('/vms', async (req, reply) => {
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const body = req.body as { name: string; image_id: string; flavor_id: string; network_id: string }
    const data = await workerService.createVm(providerId, body)
    return reply.send(data)
  })

  server.delete('/vms/:id', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.deleteVm(providerId, id)
    return reply.send(data)
  })

  server.post('/vms/:id/start', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.startVm(providerId, id)
    return reply.send(data)
  })

  server.post('/vms/:id/stop', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.stopVm(providerId, id)
    return reply.send(data)
  })

  server.post('/vms/:id/reboot', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.rebootVm(providerId, id)
    return reply.send(data)
  })

  server.get('/vms/:id/volumes', async (req, reply) => {
    const { id } = req.params as { id: string }
    const { providerId } = validateQuery(providerIdQuerySchema, req.query)
    const data = await workerService.getVmVolumes(providerId, id)
    return reply.send(data)
  })
}
